MudToggleGroup: Fix MudToggleItem selection when asynchronously loaded

jimitndiaye opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Resolves #11308 by setting the selection state of MudToggleItems as they are registered

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Unit test

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

jimitndiaye committed:
Add selection tracking to MudToggleGroup component

Introduce a new private boolean field `_itemsSelected` to manage the selection state of items in the toggle group. Implement checks to call `SetItemsSelected()` when items are added and set `_itemsSelected` to true when an item is selected, ensuring proper state management.

jimitndiaye committed:
Add tests for MudToggleGroup selection behavior

This commit introduces a new test method `ToggleGroup_SetSelectedFromValuesTest_WithAsyncItems` in `ToggleGroupTests.cs`. The method verifies the functionality of the `MudToggleGroup` component across different selection modes (single, multi, toggle) with asynchronous items. It includes assertions to ensure the selected state of items, the correct UI representation, and the internal state of the toggle group aligns with the expected values based on the selection mode.

jimitndiaye committed:
Enhance selection logic in MudToggleGroup by only affecting the registering MudToggleItem rather than all items.

jimitndiaye committed:
Fix formating.
Only set selection if selected.

versile2 commented:
> any word on when this can be reviewed/merged?

Thanks for the PR, this is definitely a bug. However, I prefer keeping things separated as much as possible, would it not make more sense just to put `SetItemsSelected();` where you have your complex if statement and not add a bool and separate logic? Am I missing something here?

versile2 commented:
Good point. 

If you want to tackle it to be perfectly honest SetItemsSelected shouldn't be used at all and should be removed. 

On Register an item should be added (call a method to register, if multi call method for multi, if single/toggle call method for single/toggle
On Unregister remove from items (but not value/values)

Then toggleitemasync could use the same methods to toggle instead of rerunning setitemsselected when something matches. I should have done this last time I worked on it. It would also fix a minor bug where if you don't @bind a value it doesn't toggle (you can clearly see why in logic)

... If you don't I can approve these changes as they are the logic is sound.

jimitndiaye commented:
@versile2 Do you mean a function like this:
```

        private void ApplySelectionState(MudToggleItem<T> item)
        {
            var selected = SelectionMode == SelectionMode.MultiSelection
                ? _values.Value?.Contains(item.Value) ?? false
                : EqualityComparer<T>.Default.Equals(_value.Value, item.Value);
            if(item.Selected != selected)
            {
                item.SetSelected(selected);
            }
        }
```
This can then be called from  `Register`, `ToggleItemAsync` and the Value(s) change handlers would then become
```
            // perform Selection after user consumes Values Changed logic
            foreach (var item in _items)
            {
                ApplySelectionState(item);
            }
```
That being the case we probably wouldn't need both `OnValueChanged` and `OnValuesChanged` - just one would suffice.

versile2 commented:
Yes, one thing to note the OnValueChanged and OnValuesChanged are only fired when the values are updated from the dev side. (ChangeHandler in parameterstate). So if the dev does `@bind-Value` and then changes the value, it will trigger the OnValueChanged so our component knows to "select" said value. And our change handler doesn't need to update the value since that's already done just update the selection. In most cases the change handler just needs to update state so the dev doesn't have to if the control doesn't recognize a change or update logic (like setting item.Selected) since it's not a 1:1 to _value.

jimitndiaye committed:
Refactor selection state management in MudToggleGroup

Removed `SetItemsSelected` method and replaced it with `ApplySelectionState` to streamline selection logic for toggle items. The new method handles both individual and bulk selection states, enhancing clarity and maintainability. The `_itemsSelected` field has been eliminated, reflecting a revised approach to tracking selection state.

jimitndiaye commented:
> Yes, one thing to note the OnValueChanged and OnValuesChanged are only fired when the values are updated from the dev side. (ChangeHandler in parameterstate). So if the dev does `@bind-Value` and then changes the value, it will trigger the OnValueChanged so our component knows to "select" said value. And our change handler doesn't need to update the value since that's already done just update the selection. In most cases the change handler just needs to update state so the dev doesn't have to if the control doesn't recognize a change or update logic (like setting item.Selected) since it's not a 1:1 to _value.

Done. I left the two separate change handlers `OnValueChanged` and `OnValuesChanged` but really they could just be eliminated and have the the change handler directly call `ApplySelectionState`.

jimitndiaye committed:
Make ApplySelectionState method overloads adjacent

jimitndiaye committed:
Fix whitespace formatting.

jimitndiaye committed:
Minor whitespace changes

jimitndiaye commented:
@versile2 The following tests from the Analyzers keep failing even though they all pass when run locally: `AllowCustomListAttributes`, `AllowDataAndAriaAttributes`, `AllowLowerCaseAttributes`, `AllowNoAttributes`. None of my changes have anything to do with them in any case. Do you have any idea what's going on?

ScarletKuro commented:
> Do you have any idea what's going on?

That's problem with the 9.0.300 SDK with the CI is fetching, similar problem we had in past https://github.com/dotnet/roslyn/issues/77255#issuecomment-2880876172
ignore for now

jimitndiaye commented:
> > Do you have any idea what's going on?
> 
> That's problem with the 9.0.300 SDK witch the CI is fetching, similar problem we had in past [dotnet/roslyn#77255 (comment)](https://github.com/dotnet/roslyn/issues/77255#issuecomment-2880876172) ignore for now
Alright. Otherwise, do the changes I made match what you had in mind?

versile2 commented:
> > ignore for now
> 
> Alright. Otherwise, do the changes I made match what you had in mind, @versile2?

This is not optimized but this works. Test it some and add a test to prevent a future mishap if you have time?

```cs
        protected internal async Task ToggleItemAsync(MudToggleItem<T> item)
        {
            var itemValue = item.Value;
            var prevItem = _items.FirstOrDefault(x => EqualityComparer<T>.Default.Equals(x.Value, _value.Value));
            if (SelectionMode == SelectionMode.MultiSelection)
            {
                var selectedValues = new HashSet<T?>(_values.Value ?? []);

                if (!selectedValues.Remove(itemValue))
                {
                    selectedValues.Add(itemValue);
                }

                await _values.SetValueAsync(selectedValues);
                prevItem = null;
            }
            else if (SelectionMode == SelectionMode.ToggleSelection)
            {
                if (EqualityComparer<T>.Default.Equals(_value.Value, itemValue))
                {
                    await _value.SetValueAsync(default); // unselect
                }
                else
                {
                    await _value.SetValueAsync(itemValue); // select
                }
            }
            else // SingleSelection
            {
                await _value.SetValueAsync(itemValue); // select
            }

            // Unselect previous
            if (prevItem is not null)
                ApplySelectionState(prevItem);

            ApplySelectionState(item);
        }
     ```

versile2 commented:
Also, Thank you for fixing something that was on my list to fix 👍

jimitndiaye commented:
Ah, that's on oversight on my part - deselecting the previous item if single/toggle selection.

jimitndiaye committed:
Add toggle item unselection test and update logic

Introduce `ToggleGroup_UnselectPreviousValue_OnToggle_Test` to verify unselection behavior in `MudToggleGroup`. Update selection logic in `MudToggleGroup.razor.cs` to ensure previously selected items are unselected based on the selection mode.

jimitndiaye committed:
Fix whitespace formatting

jimitndiaye committed:
Merge branch 'dev' into toggleitem-selection

versile2 commented on src/MudBlazor/Components/Toggle/MudToggleGroup.razor.cs:
```diff
@@ -256,20 +257,20 @@ protected override void OnAfterRender(bool firstRender)
             if (firstRender)
             {
-                SetItemsSelected();
+                ApplySelectionState();
```
Since Register and Unregister are executed at the MudToggleItem level this is not necessary

versile2 commented:
Works great, test is good. You can remove the entire OnAfterRender section as it's not necessary now. 

Also the OnValueChanged and OnValuesChanged if you want to streamline more for effeciency. Rather than run the method on all items (especially with a larger list) you could basically execute the Toggle method without updating values. Then remove the empty ApplySelectionState() method. You can get values from ParameterChangedEventArgs. I don't expect this is much of an improvement but it would be some.

```cs
        private void OnValueChanged(ParameterChangedEventArgs<T?> args)
        {
            // perform Selection after user consumes Value Changed logic
            ApplySelectionState();
        }

        private void OnValuesChanged(ParameterChangedEventArgs<IEnumerable<T?>?> args)
        {
            // perform Selection after user consumes Values Changed logic
            ApplySelectionState();
        }

jimitndiaye commented:
> Also the OnValueChanged and OnValuesChanged if you want to streamline more for effeciency. Rather than run the method on all items (especially with a larger list) you could basically execute the Toggle method without updating values. Then remove the empty ApplySelectionState() method. You can get values from ParameterChangedEventArgs. I don't expect this is much of an improvement but it would be some.

The `ApplySelectionState` method only calls `item.SetSelected` if the current selection state doesn't already match the desired selection state. Thus looping through all items would only cause the required items to be changed, not all of them.

jimitndiaye committed:
Remove OnAfterRender method from MudToggleGroup

Eliminate the OnAfterRender method in MudToggleGroup.razor.cs, which previously handled post-render actions, including applying the selection state on the first render. This change indicates that such actions are no longer necessary for the component.

ScarletKuro committed:
Merge branch 'dev' into toggleitem-selection

jimitndiaye committed:
Merge branch 'dev' into toggleitem-selection

jimitndiaye commented:
Fixed merge conflict

danielchalmers commented:
Thanks!
